src/Workspaces/Core/Portable/SemanticModelReuse/SemanticModelWorkspaceServiceFactory.SemanticModelWorkspaceService.cs

```diff
@@ -0,0 +1,175 @@
+            {
+                var linkedIds = document.GetLinkedDocumentIds();
+
+                // Get the current top level version for this document's project.  If it has changed, then we cannot
+                // reuse any existing cached data for it.  This also ensures that we can do things like find the same
+                // method body node prior to an edit just by counting it's top-level index in the file.
+                var topLevelSemanticVersion = await document.Project.GetDependentSemanticVersionAsync(cancellationToken).ConfigureAwait(false);
+
+                // If we are able to reuse a semantic model, then ensure that this is now the semantic model we're now
+                // pointing at for this document.
+                var reuseInfo = await TryReuseCachedSemanticModelAsync(
+                    map, document, bodyNode, linkedIds, topLevelSemanticVersion, cancellationToken).ConfigureAwait(false);
+                if (reuseInfo != null)
+                    return map.SetItem(document.Id, reuseInfo.Value);
+
+                // Otherwise, we couldn't reuse anything from the cache.  Return a fresh map with just the real semantic
+                // model value in it.
+                var semanticModel = await document.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+
+                var builder = ImmutableDictionary.CreateBuilder<DocumentId, SemanticModelReuseInfo?>();
+
+                // Note: we are intentionally storing the semanticmodel instance in the speculative location as well.
+                builder.Add(document.Id, new SemanticModelReuseInfo(semanticModel, semanticModel, bodyNode, topLevelSemanticVersion));
+
+                foreach (var linkedId in linkedIds)
+                    builder.Add(linkedId, null);
+
+                return builder.ToImmutable();
+            }
+
+            private static async Task<SemanticModelReuseInfo?> TryReuseCachedSemanticModelAsync(
+                ImmutableDictionary<DocumentId, SemanticModelReuseInfo?> map,
+                Document document,
+                SyntaxNode bodyNode,
+                ImmutableArray<DocumentId> linkedIds,
+                VersionStamp topLevelSemanticVersion,
+                CancellationToken cancellationToken)
+            {
+                // Get the clique (i.e. all the ids for this and the other documents it is linked to) corresponds to
+                // matches the clique we're caching.
+                using var _ = PooledHashSet<DocumentId>.GetInstance(out var documentIdClique);
+                documentIdClique.Add(document.Id);
+                documentIdClique.AddRange(linkedIds);
+
+                // if this is asking about a different doc or clique, we can't reuse anything.
+                if (!map.Keys.SetEquals(documentIdClique))
+                    return null;
```

Comment:
I would have imagined we're just looking if your document is in that map rather than comparing the entire clique. I don't know if there's funky cases where the project system is loading stuff asynchronously while you're typing code and another linked file comes on, why throw away what you have?

Reply:
> why throw away what you have?

because i don't trust what's going on in that situation, and i would prefer to be in an 'initial/clean' state.  if the doc-linking-structure changes (which should be the rare case), i don't want to screw anything up accidentally.

Put another way: the point is to optimize the mainline case as clearly as possible.  That's the bread and butter scenario for all devs.  Outside of those mainline cases, my caring drops off entirely :)

Reply:
simplified this as per yoru request.